parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

```diff
@@ -268,21 +275,36 @@ sqliteCodebase debugName root localOrRemote migrationStrategy action = do
             getShallowCausalForHash bh =
               V2Branch.hoistCausalBranch runTransaction <$> runTransaction (Ops.expectCausalBranchByCausalHash bh)
 
-            getRootBranch :: TVar (Maybe (Sqlite.DataVersion, Branch Sqlite.Transaction)) -> m (Branch m)
-            getRootBranch rootBranchCache =
-              Branch.transform runTransaction <$> runTransaction (CodebaseOps.getRootBranch branchCache getDeclType rootBranchCache)
+            getRootBranch :: m (Branch m)
+            getRootBranch =
+              Branch.transform runTransaction
+                <$> fetchRootBranch
+                  rootBranchCache
+                  (runTransaction (CodebaseOps.uncachedLoadRootBranch branchCache getDeclType))
 
-            putRootBranch :: TVar (Maybe (Sqlite.DataVersion, Branch Sqlite.Transaction)) -> Text -> Branch m -> m ()
-            putRootBranch rootBranchCache reason branch1 = do
+            putRootBranch :: Text -> Branch m -> m ()
+            putRootBranch reason branch1 = do
               now <- liftIO getCurrentTime
               withRunInIO \runInIO -> do
-                runInIO do
-                  runTransaction do
-                    let emptyCausalHash = Cv.causalHash1to2 $ Branch.headHash Branch.empty
-                    fromRootCausalHash <- fromMaybe emptyCausalHash <$> Ops.loadRootCausalHash
-                    let toRootCausalHash = Cv.causalHash1to2 $ Branch.headHash branch1
-                    CodebaseOps.putRootBranch rootBranchCache (Branch.transform (Sqlite.unsafeIO . runInIO) branch1)
-                    Ops.appendReflog (Reflog.Entry {time = now, fromRootCausalHash, toRootCausalHash, reason})
+                -- this is naughty, the type says Transaction but it
+                -- won't run automatically with whatever Transaction
+                -- it is composed into unless the enclosing
+                -- Transaction is applied to the same db connection.
```

Comment:
Yeah, now that we have better ways to run Transactions outside of the codebase itself (with Codebase.runTransaction) it'd probably be better to return `Branch Transaction` here, but I think that'd be a big refactor.

What if instead of using unliftIO we  did `Branch.transform runTransaction` here? Then it'd make a new transaction for each layer of the branch.

Reply:
I agree that it would be better to return `Branch Transaction`. This PR attempts to preserve the existing behavior except for fixing the cache bugs/adding the file lock.

What advantages are you thinking of for starting a new transaction for each layer? Avoiding a long-running transaction would be good.

Fetching layers in separate transactions (as `Branch Transaction` or `Branch IO`) is spooky. We need to be sure to never return a `Branch` with layers from distinct snapshots, as that could return a branch that not only doesn't correspond to the current db state, but doesn't correspond to any db state.

Reply:
> What advantages are you thinking of for starting a new transaction for each layer? Avoiding a long-running transaction would be good.

Yeah, it'd be avoiding a long-running transaction, also I'm not sure how exactly it's currently wired up, but is that transaction ever closed currently?

> We need to be sure to never return a Branch with layers from distinct snapshots

True, but I don't believe this is actually possible, since each "layer"'s transactions/IO is pre-seeded with which hash it'll load when forced, and the contents of those hashes are immutable.

So I don't think there's any real benefit to keeping them all in a single transaction, and doing so locks up that SQLite connection; when is it actually released? Does that transaction stay open until the codebase is closed?